### ISSUE TITLE
Crash if email address of contacts is malformed

### ISSUE THREAD

**Issue by sim6** [thunderbird/thunderbird-android#1889]
### Expected behavior
Maybe a warning.

### Actual behavior
Crash.

### Steps to reproduce
1. Create a contact email with email "user@domain.tld>", "<user@domain.tld" or "us<er@domain.tld"... with or without quotes.
2. Touch the email from the contacts application.
3. Crash.

### Environment
K-9 Mail version:
5.200

Android version:
5.1

Account type (IMAP, POP3, WebDAV/Exchange):
IMAP

**Comment by Valodim** [comment]
This is possibly caused by #1874. Please see if this persists in version 5.201

**Comment by sim6** [comment]
OK, thanks.

**Comment by sim6** [comment]
The crash persists in 5.202.

**Comment by philipwhiuk** [comment]
Reproduced. Same sort of bug, but different bit of code. Simple to fix.

```
 java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.String.equals(java.lang.Object)' on a null object reference
     at com.fsck.k9.mail.Address.equals(Address.java:170)
     at com.fsck.k9.view.RecipientSelectView$Recipient.equals(RecipientSelectView.java:620)
     at java.util.ArrayList.indexOf(ArrayList.java:305)
     at java.util.ArrayList.contains(ArrayList.java:288)
     at com.tokenautocomplete.TokenCompleteTextView.insertSpan(TokenCompleteTextView.java:1060)
     at com.tokenautocomplete.TokenCompleteTextView.access$700(TokenCompleteTextView.java:54)
     at com.tokenautocomplete.TokenCompleteTextView$3.run(TokenCompleteTextView.java:924)
     at android.os.Handler.handleCallback(Handler.java:751)
     at android.os.Handler.dispatchMessage(Handler.java:95)
     at android.os.Looper.loop(Looper.java:154)
     at android.app.ActivityThread.main(ActivityThread.java:6119)
     at java.lang.reflect.Method.invoke(Native Method)
     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
```

**Comment by philipwhiuk** [comment]
There's some stuff to think about here. Making `Address.equals` handle null mAddress is easy. Dealing with an Address without an actual email is harder. Right now it gets loaded into the email view fine. But then if you click on that Token it will crash like this:

```
java.lang.RuntimeException: An error occurred while executing doInBackground()
    at android.os.AsyncTask$3.done(AsyncTask.java:325)
    at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
    at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
    at java.util.concurrent.FutureTask.run(FutureTask.java:242)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
    at java.lang.Thread.run(Thread.java:761)
 Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'int java.lang.String.length()' on a null object reference
    at android.os.Parcel.readException(Parcel.java:1690)
    at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:183)
    at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:135)
    at android.content.ContentProviderProxy.query(ContentProviderNative.java:421)
    at android.content.ContentResolver.query(ContentResolver.java:532)
    at android.content.ContentResolver.query(ContentResolver.java:474)
    at com.fsck.k9.activity.compose.RecipientLoader.fillCryptoStatusData(RecipientLoader.java:263)
    at com.fsck.k9.activity.compose.RecipientLoader.loadInBackground(RecipientLoader.java:122)
    at com.fsck.k9.activity.compose.RecipientLoader.loadInBackground(RecipientLoader.java:25)
    at android.content.AsyncTaskLoader.onLoadInBackground(AsyncTaskLoader.java:312)
    at android.content.AsyncTaskLoader$LoadTask.doInBackground(AsyncTaskLoader.java:69)
    at android.content.AsyncTaskLoader$LoadTask.doInBackground(AsyncTaskLoader.java:66)
    at android.os.AsyncTask$2.call(AsyncTask.java:305)
    at java.util.concurrent.FutureTask.run(FutureTask.java:237)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133) 
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607) 
    at java.lang.Thread.run(Thread.java:761) 
```

The reason is that (working outwards):

1. OpenKeychain doesn't handle a null element in the selectionArgs array: https://github.com/open-keychain/open-keychain/blob/v4.1/OpenKeychain/src/main/java/org/sufficientlysecure/keychain/remote/KeychainExternalProvider.java#L188
2. We assume all recipient addresses have emails: https://github.com/k9mail/k-9/blob/master/k9mail/src/main/java/com/fsck/k9/activity/compose/RecipientLoader.java#L128 when we add them to the 'map' (i.e. `address.getAddress()` is assumed to not be null)

So maybe that for-loop just wants a null-check.

On to actually sending:

It ought to result in some error. It doesn't. What actually happens is it sits in the outbox forever. The only oddball thing is that the message 'No headers available for this message!' is printed at Debug.

TLDR: This needs work.

**Comment by sim6** [comment]
Not fixed for 5.203

**Comment by philipwhiuk** [comment]
The change to fix this wasn't merged until a couple of days ago, it should be in the next release.

**Comment by sim6** [comment]
Thanks, I will try it.

**Comment by sim6** [comment]
Thanks, not reproducibe in v5.205. I cannot test v5.204 because it isn't in f-droid.
